python-ortools-region-allocation/main.py

```diff
@@ -0,0 +1,163 @@
+    for region in input["regions"]:
+        solver.Add(sum(assignments[(region["id"], hub["id"])] for hub in input["hubs"]) == 1)
+
+    # Make sure that the demand of the regions assigned to the hub is covered by its capacity.
+    for hub in input["hubs"]:
+        solver.Add(
+            sum(assignments[(region["id"], hub["id"])] * region["demand"] for region in input["regions"])
+            <= hub["capacity"]
+        )
+
+    # Set the objective function to minimize the total distance.
+    objective = solver.Objective()
+    for region in input["regions"]:
+        for hub in input["hubs"]:
+            distance = distance_matrix[(region["id"], hub["id"])]
+            objective.SetCoefficient(assignments[(region["id"], hub["id"])], distance)
+    objective.SetMinimization()
+
+    # Solves the problem.
+    status = solver.Solve()
+
+    # Get the assigned regions.
+    assignments_result = {
+        region["id"]: next(
+            hub["id"] for hub in input["hubs"] if assignments[(region["id"], hub["id"])].solution_value() > 0.5
+        )
+        for region in input["regions"]
+    }
+    solution = {
+        "assignments": assignments_result,
+        "hubs": input["hubs"],
+        "regions": input["regions"],
+    }
+
+    # Collect some statistics.
+    statistics = {
+        "run": {
+            "duration": time.time() - start_time,
+        },
+        "result": {
+            "duration": solver.WallTime() / 1000,
+            "value": solver.Objective().Value(),
+            "custom": {
+                "status": STATUS.get(status, "unknown"),
+                "variables": solver.NumVariables(),
+                "constraints": solver.NumConstraints(),
+            },
+        },
+    }
+
+    return solution, statistics
+
+
+def calculate_distance(region: dict, hub: dict) -> float:
+    """Calculates the distance between a region and a hub using the haversine formula."""
+    region_coords = region["center"]
+    hub_coords = hub["center"]
+    lat1, lon1 = region_coords["y"], region_coords["x"]
+    lat2, lon2 = hub_coords["y"], hub_coords["x"]
+
+    # Convert latitude and longitude from degrees to radians
+    lat1, lon1, lat2, lon2 = map(math.radians, [lat1, lon1, lat2, lon2])
+
+    # Haversine formula
+    dlat = lat2 - lat1
+    dlon = lon2 - lon1
+    a = math.sin(dlat / 2) ** 2 + math.cos(lat1) * math.cos(lat2) * math.sin(dlon / 2) ** 2
+    c = 2 * math.asin(math.sqrt(a))
+
+    # Radius of Earth in kilometers. Use 3956 for miles
+    r = 6371
+
+    # Calculate the result
+    return c * r
+
+
+class DecisionModel(nextmv.Model):
+    def solve(self, input: nextmv.Input) -> nextmv.Output:
+        """Solves the given problem and returns the solution."""
+
+        # Redirect solver chatter to stderr.
+        nextmv.redirect_stdout()
+
+        # Solve the problem.
+        solution, statistics = solve(input.data, input.options.provider, input.options.duration)
```

Comment:
Why not add all the code from `solve` in this `DecisionModel.solve` func?

Reply:
Yeah, this was just for the demo / tech talk today. Good point. I will restructure it now that it is its own app. :blush: